// petri-specs/test/testkit.js
function spec(name, fn = x => x) {
  return {
    [name]: fn({
      scopes: ['my-scope'],
      owner: 'shahart@wix.com',
      onlyForLoggedInUsers: true,
      controlGroup: '',
      variants: []
    })
  };
}

function multipleSpecs(names, fn = x => x) {
  const specsMap = {};
  names.forEach(specName => Object.assign(specsMap, spec(specName, fn)));
  return specsMap;
}

function testGroupsSpec(name, fn = x => x) {
  return {
    [name]: fn({
      scopes: ['my-scope'],
      owner: 'shahart@wix.com',
      onlyForLoggedInUsers: true,
      testGroups: []
    })
  };
}

function singleScopeSpec(name, fn = x => x) {
  return {
    [name]: fn({
      scope: 'my-scope',
      owner: 'shahart@wix.com',
      onlyForLoggedInUsers: true,
      testGroups: []
    })
  };
}

function multipleSingleScopeSpecs(names, fn = x => x) {
  const specsMap = {};
  names.forEach(specName => Object.assign(specsMap, singleScopeSpec(specName, fn)));
  return specsMap;
}

function translationSpec(name, fn = x => x) {
  const specs = spec(name, fn);
  specs[name].controlGroup = '0';
  specs[name].variants = ['1', '2'];

  return specs;
}

function pom(params = {}) {
  const groupId = params.groupId || 'com.wixpress';
  const scope = params.artifactId || 'my-scope';
  const owner = params.owner || 'Ernie';
  const email = params.email || 'shahart@wix.com';
  const version = params.version || '1.0.0';

  return `<?xml version="1.0" encoding="UTF-8"?>
  <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
      <modelVersion>4.0.0</modelVersion>
      <groupId>${groupId}</groupId>
      <artifactId>${scope}</artifactId>
      <version>${version}-SNAPSHOT</version>
      <name>${scope}</name>
      <description>${scope}</description>
      <packaging>pom</packaging>

      <parent>
          <groupId>com.wixpress.common</groupId>
          <artifactId>wix-master-parent</artifactId>
          <version>100.0.0-SNAPSHOT</version>
      </parent>

      <developers>
          <developer>
              <name>${owner}</name>
              <email>${email}</email>
              <roles>
                  <role>owner</role>
              </roles>
          </developer>
      </developers>
  </project>
  `;
}

function baseFsWith(files = {}) {
  return Object.assign({
    'package.json': '',
    'pom.xml': pom()
  }, files);
}

function translationWithSpecs(...specs) {
  const translations = {};
  specs.forEach(spec => {
    translations[spec] = 'Hello world';
    translations[`${spec}|abTranslate_1`] = 'Hi world';
    translations[`${spec}|abTranslate_2`] = 'Yo world';
  });

  return translations;
}

module.exports = {
  baseFsWith,
  spec,
  multipleSpecs,
  testGroupsSpec,
  singleScopeSpec,
  multipleSingleScopeSpecs,
  translationSpec,
  pom,
  translationWithSpecs
};
