// ab-translate/dist/test/abTranslate.driver.js
'use strict';

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

jest.mock('glob'); // eslint-disable-line no-undef
jest.mock('fs'); // eslint-disable-line no-undef

var ABTranslateDriver = function ABTranslateDriver() {
  var _this = this;

  (0, _classCallCheck3.default)(this, ABTranslateDriver);

  this.glob = require('glob');
  this.fs = require('fs');
  this.glob.__clearFiles();
  this.fs.__clearFiles();

  this.given = {
    pomIsDefined: function pomIsDefined() {
      _this.fs.__setFile('pom.xml', _this.POM_XML);
      return _this;
    },
    translationFiles: function translationFiles(files) {
      _this.glob.sync = jest.fn(function () {
        return (0, _keys2.default)(files);
      }); // eslint-disable-line no-undef
      _this.fs.__setFiles(files);
      return _this;
    }
  };

  this.get = {
    globPath: function globPath() {
      return _this.glob.sync.mock.calls[0][0];
    }
  };

  this.POM_XML = '<?xml version="1.0" encoding="UTF-8"?>\n    <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">\n      <modelVersion>4.0.0</modelVersion>\n      <groupId>com.wixpress.cx</groupId>\n      <artifactId>ab-translate</artifactId>\n      <packaging>pom</packaging>\n      <name>ab-translate</name>\n      <description>ab-translate</description>\n      <version>1.0.0-SNAPSHOT</version>\n\n      <parent>\n          <groupId>com.wixpress.common</groupId>\n          <artifactId>wix-master-parent</artifactId>\n          <version>100.0.0-SNAPSHOT</version>\n      </parent>\n\n      <developers>\n          <developer>\n              <name>Yair Haimovitch</name>\n              <email>yairh@wix.com</email>\n              <roles>\n                  <role>owner</role>\n              </roles>\n          </developer>\n      </developers>\n    </project>\n    ';
  this.FILE1 = '{"button-text": "default", "button-text|abTranslate_1": "hello","button-text|abTranslate_2": "hello 2"}';
};

module.exports = ABTranslateDriver;
